lib/sync/fullsync.go

```diff
@@ -0,0 +1,547 @@
+	}
+
+	if msg.BestBlock {
+		pv := f.peers.get(from)
+		if uint(pv.bestBlockNumber) != msg.Number {
+			repChange = &Change{
+				who: from,
+				rep: peerset.ReputationChange{
+					Value:  peerset.BadBlockAnnouncementValue,
+					Reason: peerset.BadBlockAnnouncementReason,
+				},
+			}
+			return repChange, fmt.Errorf("%w: peer %s, on handshake #%d, on announce #%d",
+				errMismatchBestBlockAnnouncement, from, pv.bestBlockNumber, msg.Number)
+		}
+	}
+
+	currentTarget := f.peers.getTarget()
+	if msg.Number >= uint(currentTarget) {
+		return nil, nil
+	}
+
+	blockAnnounceHeader := types.NewHeader(msg.ParentHash, msg.StateRoot, msg.ExtrinsicsRoot, msg.Number, msg.Digest)
+	blockAnnounceHeaderHash := blockAnnounceHeader.Hash()
+
+	logger.Infof("received block announce from %s: #%d (%s) best block: %v",
+		from,
+		blockAnnounceHeader.Number,
+		blockAnnounceHeaderHash,
+		msg.BestBlock,
+	)
+
+	// check if their best block is on an invalid chain, if it is,
+	// potentially downscore them for now, we can remove them from the syncing peers set
+	highestFinalized, err := f.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return nil, fmt.Errorf("get highest finalised header: %w", err)
+	}
+
+	if blockAnnounceHeader.Number <= highestFinalized.Number {
+		repChange = &Change{
+			who: from,
+			rep: peerset.ReputationChange{
+				Value:  peerset.BadBlockAnnouncementValue,
+				Reason: peerset.BadBlockAnnouncementReason,
+			},
+		}
+		return repChange, fmt.Errorf("%w: peer %s, block number #%d (%s)",
+			errPeerOnInvalidFork, from, blockAnnounceHeader.Number, blockAnnounceHeaderHash.String())
+	}
+
+	has, err := f.blockState.HasHeader(blockAnnounceHeaderHash)
+	if err != nil {
+		return nil, fmt.Errorf("checking if header exists: %w", err)
+	}
+
+	if !has {
+		f.unreadyBlocks.newHeader(blockAnnounceHeader)
+		request := network.NewBlockRequest(*variadic.FromHash(blockAnnounceHeaderHash),
+			1, network.RequestedDataBody+network.RequestedDataJustification, network.Ascending)
+		f.requestQueue.PushBack(request)
+	}
+
+	return nil, nil
+}
+
+type RequestResponseData struct {
+	req          *network.BlockRequestMessage
+	responseData []*types.BlockData
+}
+
+func validateResults(results []*syncTaskResult, badBlocks []string) (repChanges []Change,
+	peersToBlock []peer.ID, validRes []RequestResponseData) {
+
+	repChanges = make([]Change, 0)
+	peersToBlock = make([]peer.ID, 0)
+	validRes = make([]RequestResponseData, 0, len(results))
+
+resultLoop:
+	for _, result := range results {
+		request := result.request.(*network.BlockRequestMessage)
+
+		if !result.completed {
+			continue
+		}
+
+		response := result.response.(*network.BlockResponseMessage)
+		if request.Direction == network.Descending {
+			// reverse blocks before pre-validating and placing in ready queue
+			slices.Reverse(response.BlockData)
+		}
+
+		err := validateResponseFields(request, response.BlockData)
+		if err != nil {
+			logger.Criticalf("validating fields: %s", err)
```

Comment:
Nit/should this just be error instead of critical?

Reply:
can be a warnf